editoast/fga_migrations/src/migration.rs

```diff
@@ -0,0 +1,509 @@
+use crate::client::Migration as TargetMigration;
+use fga::client::Client;
+use fga::client::ConnectionSettings;
+use fga::client::InitializationError;
+use fga::client::QueryError;
+use fga::client::RequestFailure;
+use fga::client::TooManyTuples;
+use fga::compile_model;
+use fga::model::Relation;
+use indexmap::IndexMap;
+use indexmap::indexmap;
+use itertools::Either;
+use sha1::Digest;
+use sha1::Sha1;
+use tracing::error;
+use tracing::info;
+use tracing::instrument;
+use url::Url;
+
+const MIGRATIONS_MODEL: &str = include_str!("../migrations_model.fga");
+
+#[cfg(not(test))]
+static OPENFGA_MIGRATIONS: LazyLock<IndexMap<&'static str, &'static str>> = LazyLock::new(|| {
+    // The migrations declaration order matches the order in which they will be applied
+    indexmap! {
+        "initial_model" => include_str!("../migrations/1_initial_model.fga"),
+    }
+});
+
+#[cfg(test)]
+static OPENFGA_MIGRATIONS: LazyLock<IndexMap<&'static str, &'static str>> = LazyLock::new(|| {
+    // The migrations declaration order matches the order in which they will be applied
+    indexmap! {
+        "initial_model" => include_str!("../tests/migrations/1_initial_model.fga"),
+        "intermediate_model" => include_str!("../tests/migrations/2_intermediate_model.fga"),
+        "final_model" => include_str!("../tests/migrations/3_latest_model.fga"),
+    }
+});
+
+#[derive(Debug, Clone)]
+pub struct Settings {
+    pub migration_settings: MigrationSettings,
+    pub telemetry: TelemetrySettings,
+}
+
+#[derive(Debug, Clone)]
+pub struct MigrationSettings {
+    pub connection_settings: ConnectionSettings,
+    pub migration_store: String,
+    pub authorization_store: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct TelemetrySettings {
+    pub endpoint: Url,
+}
+
+impl From<&CliParser> for Settings {
+    fn from(client: &CliParser) -> Self {
+        let connection_settings = ConnectionSettings::new(
+            client.openfga_url.host().unwrap().to_string(),
+            client.openfga_url.port_or_known_default().unwrap(),
+            fga::client::Limits {
+                max_checks_per_batch_check: client.tuple_reads,
+                max_tuples_per_write: client.tuple_writes,
+            },
+        );
+        Settings {
+            migration_settings: MigrationSettings {
+                connection_settings,
+                migration_store: client.migrations_store.clone(),
+                authorization_store: client.authorization_store.clone(),
+            },
+            telemetry: TelemetrySettings {
+                endpoint: url::Url::parse(&client.telemetry_endpoint).expect("Invalid url"),
+            },
+        }
+    }
+}
+
+#[derive(fga::Type, fga::User, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Migration(pub usize);
+
+#[derive(fga::Type, fga::Object, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Model(pub String);
+
+fga::relations! {
+    Model {
+        apply: Migration
+    }
```

Comment:
Isn't it the other way around?

```suggestion
    Migration {
        apply: Model
    }
```

We "select a migration to apply a model"

Reply:
It would make the migration the object and the model the user. Wouldn't that be weird ? The migration should be using a model, not the other way around. Maybe we should instead rename the relation to something else like `owner` or `can_apply` ? `A model has an owner` / `A migration can apply a model`.

Maybe I didn't understand openfga user / object sematics correctly ?

In our case we mostly want to associate a migration to a model and store them somewhere in order to be able to tell what migration is currently applied. We chose to store them in an openfga store, which forces us to define them as a user and an object and to also define their relationship. I don't think openfga semantics add much in practice though: using them is a necessity if we want to handle migrations by interacting only with the openfga server (and not with a postgreSQL database for example), but reverting the relationship wouldn't change much in practice: the workflow would still be to retrieve both the current migration and model that are stored somewhere and check how they match with the target migration of the current migration run.

For those reasons, I didn't give the many thoughts to the nature of the relationship between the model types and used the model defined in the related issue without challenging it. In the end, I don't think it changes much apart from implementation details (the queries to send to retrieve both the migration and the model will be slightly different).

What do you think ?

Reply:
the model will change